maskit/_masked_circuits.py

```diff
@@ -0,0 +1,475 @@
+        Note that this mask is readonly.
+        """
+        base = super().mask
+        if Axis.PARAMETERS in self.freeze_masks:
+            base[self.freeze_masks[Axis.PARAMETERS].mask] = True
+        if Axis.LAYERS in self.freeze_masks:
+            base[self.freeze_masks[Axis.LAYERS].mask, :] = True
+        if Axis.WIRES in self.freeze_masks:
+            base[:, self.freeze_masks[Axis.WIRES].mask] = True
+        return base
+
+    def freeze_mask_for_axis(self, axis: Axis) -> Mask:
+        return self.freeze_masks[axis]
+
+    def freeze(
+        self,
+        axis: Axis = Axis.LAYERS,
+        amount: Optional[Union[int, float]] = None,
+        mode: Mode = Mode.ADD,
+    ):
+        """
+        Freezes the parameter values for a given ``axis`` that is of type
+        :py:class:`~.Axis`. The freezing is applied ``amount``times
+        and depends on the given ``mode`` of type :py:class:`~.PerturbationMode`.
+        If no amount is given, that is ``amount=None``, a random ``amount`` is
+        determined given by the actual size of the py:attr:`~.mask`. The ``amount``
+        is automatically limited to the actual size of the py:attr:`~.mask`.
+
+        :param amount: Number of items to freeze, defaults to None
+        :param axis: Which mask to freeze, defaults to PerturbationAxis.LAYERS
+        :param mode: How to freeze, defaults to PerturbationMode.ADD
+        :raises NotImplementedError: Raised in case of an unknown mode
+        """
+        assert mode in list(
+            Mode
+        ), f"The selected perturbation mode {mode} is not supported."
+        if amount == 0:
+            return
+        if axis in self.freeze_masks:
+            self.freeze_mask_for_axis(axis).perturb(amount=amount, mode=mode)
+        else:
+            raise NotImplementedError(f"The perturbation {axis} is not supported")
+
+    def copy(self: Self) -> Self:
+        """Returns a copy of the current FreezableMaskedCircuit."""
+        clone = super().copy()
+        clone.freeze_masks = {}
+        for axis, mask in self.freeze_masks.items():
+            clone.freeze_masks[axis] = mask.copy()
+        return clone
+
+    @classmethod
+    def full_circuit(
+        cls,
+        parameters: np.ndarray,
+        layers: int,
+        wires: int,
+        dynamic_parameters: bool = True,
+        default_value: Optional[float] = None,
+        entangling_mask: Optional[Mask] = None,
+        wire_mask: Optional[Mask] = None,
+        layer_mask: Optional[Mask] = None,
+        parameter_mask: Optional[Mask] = None,
+    ):
+        initializable_masks = [axis for axis in Axis if axis != Axis.ENTANGLING]
+        if wire_mask is not None:
+            initializable_masks.remove(Axis.WIRES)
+        if layer_mask is not None:
+            initializable_masks.remove(Axis.LAYERS)
+        if parameter_mask is not None:
+            initializable_masks.remove(Axis.PARAMETERS)
```

Comment:
Might want to use a loop here, since it's basically the same thing again and again. *Roughly* like this (untested, may contain typos):
```suggestion
        for axis, mask in ((Axis.WIRES, wire_mask), (Axis.LAYERS, layer_mask),(Axis.PARAMETERS, parameter_mask)):
            if mask is not None:
                initializable_masks.remove(axis)
```
This likely won't make it shorter, but makes it clear that it's just doing the same thing.